corgi/tasks/prod_defs.py

```diff
@@ -263,10 +265,34 @@ def parse_product_stream(
             "obj": product_stream,
         },
     )
+    _clean_orphaned_brew_tags(set(brew_tags_dict.keys()), name, str(product_stream.pk))
     parse_errata_info(errata_info, product, product_stream, product_stream_node, product_version)
     product_stream.save_product_taxonomy()
 
 
+def _clean_orphaned_brew_tags(brew_tags: set[str], name: str, product_stream_pk: str):
+    existing_stream_relation_tags = (
+        ProductComponentRelation.objects.filter(
+            type=ProductComponentRelation.Type.BREW_TAG, product_ref=name
+        )
+        .values_list("external_system_id", flat=True)
+        .distinct()
+    )
+    removed_tags = set(existing_stream_relation_tags) - brew_tags
+    for tag in removed_tags:
+        logger.info(f"Removing tag {tag} from stream {name}")
+        for brew_tag_relation in ProductComponentRelation.objects.filter(
+            type=ProductComponentRelation.Type.BREW_TAG, product_ref=name, external_system_id=tag
+        ).iterator():
+            if brew_tag_relation.software_build_id:
```

Comment:
I think you should be able to just filter the queryset directly, instead of fetching all the relations from the DB and checking them using an if statement:
```suggestion
        for brew_tag_relation in ProductComponentRelation.objects.filter(
            type=ProductComponentRelation.Type.BREW_TAG, product_ref=name, external_system_id=tag, software_build_id__isnull=False
        ).iterator():
```

Reply:
The problem with that is we also want to delete any unprocessed relations, where software_build_id__isnull=True. But if the SoftwareBuild isn't linked as a foreign key that means it's been processed and we want to remove it from the build. I'll add a comment.